src/main/java/com/multithreading/TextProcessor.java

```diff
@@ -1,38 +1,67 @@
 package com.multithreading;
 
+import java.util.ArrayList;
+import java.util.List;
+
 class Capitalize implements Runnable {
-    private final char[] chars;
-    private final int index;
+    private volatile boolean isRunning = true;
+    private char character;
 
-    public Capitalize(char[] chars, int index) {
-        this.chars = chars;
-        this.index = index;
+    public Capitalize(char character) {
+        this.character = character;
     }
 
     @Override
     public void run() {
-        synchronized (chars) {
-            chars[index] = Character.toUpperCase(chars[index]);
+        if (isRunning) {
+            this.character = Character.toUpperCase(character);
         }
     }
+
+    public void stop() {
+        this.isRunning = false;
+    }
+
+    public char getChar() {
+        return character;
+    }
 }
 
 public class TextProcessor {
 
+    private final List<Thread> threads;
+    private final List<Capitalize> capitalizeTasks;
+
+    TextProcessor() {
+        this.capitalizeTasks = new ArrayList<>();
+        this.threads = new ArrayList<>();
+    }
+
     public String process(String input) throws InterruptedException {
-        Thread[] threads = new Thread[input.length()];
+        int inputLength = input.length();
 
         char[] chars = input.toCharArray();
+        char[] capitalizedChars = new char[inputLength];
 
-        for (int i = 0; i < chars.length; i++) {
-            threads[i] = new Thread(new Capitalize(chars, i));
-            threads[i].start();
+        for (int i = 0; i < inputLength; i++) {
+            capitalizeTasks.add(new Capitalize(chars[i]));
```

Comment:
note: because you didn't clear the tasks array before process is called you might not end up in sync between the input index and the task index if you call process more than once so your joining will fail.